oauth2/handler.go

```diff
@@ -1154,3 +1207,175 @@ func (h *Handler) logOrAudit(err error, r *http.Request) {
+// https://openid.net/specs/openid-connect-userinfo-vc-1_0.html.
+//
+//	Consumes:
+//	- application/json
+//
+//	Schemes: http, https
+//
+//	Responses:
+//	  200: verifiableCredentialResponse
+//	  400: verifiableCredentialPrimingResponse
+//	  default: errorOAuth2
+func (h *Handler) createVerifiableCredential(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	session := NewSessionWithCustomClaims("", h.c.AllowedTopLevelClaims(ctx))
+	accessToken := fosite.AccessTokenFromRequest(r)
+	tokenType, _, err := h.r.OAuth2Provider().IntrospectToken(ctx, accessToken, fosite.AccessToken, session)
+
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(err))
+		return
+	}
+	if tokenType != fosite.AccessToken {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("the token is not an access token")))
+		return
+	}
+
+	var request CreateVerifiableCredentialRequestBody
+	if err := json.NewDecoder(r.Body).Decode(&request); err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(err))
+		return
+	}
+
+	if request.Format != "jwt_vc_json" {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("the format is not supported")))
+		return
+	}
+	if request.Proof == nil {
+		// Handle priming request
+		nonceLifespan := h.r.Config().GetVerifiableCredentialsNonceLifespan(ctx)
+		nonceExpiresIn := time.Now().Add(nonceLifespan).UTC()
+		nonce, err := h.r.OAuth2Storage().NewNonce(ctx, accessToken, nonceExpiresIn)
+		if err != nil {
+			h.r.Writer().WriteError(w, r, errorsx.WithStack(err))
+			return
+		}
+		w.WriteHeader(http.StatusBadRequest)
+		h.r.Writer().Write(w, r, &VerifiableCredentialPrimingResponse{
+			RFC6749ErrorJson: fosite.RFC6749ErrorJson{
+				Name:        "missing_proof",
+				Description: "Could not issue a verifiable credential because the proof is missing in the request.",
+			},
+			Format:          "jwt_vc",
+			Nonce:           nonce,
+			NonceExperiesIn: int64(nonceLifespan.Seconds()),
+		})
+		return
+	}
+	if request.Proof.ProofType != "jwt" {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("the proof type is not supported")))
+		return
+	}
+
+	header, _, ok := strings.Cut(request.Proof.JWT, ".")
+	if !ok {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("invalid JWT in proof")))
+		return
+	}
+
+	rawHeader, err := base64.RawURLEncoding.DecodeString(header)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("invalid JWT header in proof")))
+		return
+	}
+	jwk := gjson.GetBytes(rawHeader, "jwk").String()
+	proofJWK, err := josex.LoadJSONWebKey([]byte(jwk), true)
+	if err != nil {
+		h.r.Writer().WriteError(w, r, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("invalid JWK in JWT header in proof")))
+		return
+	}
+
+	token, err := jwt.Parse(request.Proof.JWT, func(token *jwt.Token) (any, error) {
```

Comment:
I'm pretty sure that you can provide a custom key discovery function to the JWT library, which would be way more elegant and secure (and has lots of tests)

Reply:
I do pass the discovery function here. The key is discovered from the JWT header.